### PR TITLE
Add listeners by creation time and cleanup error reports for conflicts

### PR DESCRIPTION
**Pull request by nmittler**
This PR adds the CreationTime to the service struct and populates it
in the k8s service registry. When available, services will be sorted
by their creation time. Conflicts will cause the listeners for newer
services to be dropped.

Partial fix for #5992

### CODE REVIEW

**Comment by nmittler** [comment]
This rebases #6993 onto the `release-1.0` branch.

File: istioctl/cmd/istioctl/istioctl_test.go [hunk 1]
```diff
@@ -46,16 +46,6 @@ type testCase struct {
 	wantException bool
 }
 
-type testCaseGoldenFile struct {
```
Comment (esnible) [inline istioctl/cmd/istioctl/istioctl_test.go]: Good catch!

File: istioctl/cmd/istioctl/istioctl_test.go [hunk 2]
```diff
@@ -189,14 +179,14 @@ func TestGet(t *testing.T) {
 			configs: testGateways,
 			args:    strings.Split("get gateways -n default", " "),
 			expectedOutput: `GATEWAY NAME       HOSTS     NAMESPACE   AGE
-bookinfo-gateway   *         default     <unknown>
```
Comment (esnible) [inline istioctl/cmd/istioctl/istioctl_test.go]: Nit: Would it ever be possible that this test might fail if it takes >1 second to execute?  I'm concerned that "0s" might turn into "1s".

Comment (nmittler) [reply istioctl/cmd/istioctl/istioctl_test.go]: Yeah it would fail in that case.  Should be *very* unlikely, though.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 3]
```diff
@@ -337,6 +337,46 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
...
+	listener *xdsapi.Listener
+}
+
+func getServiceName(env model.Environment, service *model.Service) string {
+	attrs, err := env.GetServiceAttributes(service.Hostname)
+	if err != nil || attrs == nil || attrs.Name == "" {
+		return "unknown"
+	}
+	return fmt.Sprintf("%s/%s", attrs.Namespace, attrs.Name)
+}
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: why do you need this? why isn't hostname sufficient?
This is also going to fail when you have a k8s service and a non k8s service hostname, to the extent that the error message you throw will be very confusing: "conflict with foo/bar and unknown". 

The service Hostname has all the elements needed to uniquely distinguish the service (foo.bar.svc.cluster.local). I don't think the extra name/namespace will add any additional information.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 4]
```diff
@@ -337,6 +337,46 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
...
+
+func getClusterName(service *model.Service, servicePort *model.Port) string {
+	return model.BuildSubsetKey(model.TrafficDirectionOutbound, "",
+		service.Hostname, servicePort.Port)
+}
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: This also seems like an unnecessary function just for the printing convenience. Given that its just two more args, and called/used only once in this entire file.

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 3 above)
Comment (costinm) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: I think this is also expensive ( env.GetServiceAttributes). 

Instead the namespace + name should be populated at conversion time, where they are available.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 5]
```diff
@@ -337,6 +337,46 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
...
+
+func handleOutboundListenerConflict(env model.Environment, service *model.Service, servicePort *model.Port, current listenerEntry, listenerMapKey string) {
+	conflictingOutbound.Add(1)
+	log.Warnf("buildSidecarOutboundListener: Omitting listener for service '%s' due to conflict with service '%s'. "+
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: This is likely to go away and be replaced by the metric, this kind of log is hard to track in pilot logs due to noise (each log is repeated for every single sidecar, along with lots of other logs )

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 6]
```diff
@@ -354,6 +394,9 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
 func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env model.Environment, node model.Proxy,
 	proxyInstances []*model.ServiceInstance, services []*model.Service) []*xdsapi.Listener {
 
+	// Sort the services in order of creation.
+	services = sortServicesByCreationTime(services)
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: Just want to be sure that it works for services that are not loaded from k8s - i.e. we have some default value for time, can be the timestamp of load.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 7]
```diff
@@ -493,16 +532,15 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env model.En
...
-		if err := l.Validate(); err != nil {
-			log.Warnf("buildSidecarOutboundListeners: error validating listener %s (type %v): %v", name, ltype, err)
+		if err := l.listener.Validate(); err != nil {
+			log.Warnf("buildSidecarOutboundListeners: error validating listener %s (type %v): %v", name, l.servicePort.Protocol, err)
 			invalidOutboundListeners.Add(1)
 			continue
 		}
-		if ltype.IsTCP() {
-			tcpListeners = append(tcpListeners, l)
+		if l.servicePort.Protocol.IsTCP() {
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: Was this change needed ? Shriram - do you remember why we had listenerTypeMap ?

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 8]
```diff
@@ -364,13 +407,9 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env model.En
 
 	var tcpListeners, httpListeners []*xdsapi.Listener
 	var currentListener *xdsapi.Listener
-	listenerTypeMap := make(map[string]model.Protocol)
-	listenerMap := make(map[string]*xdsapi.Listener)
+	listenerMap := make(map[string]listenerEntry)
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: can we use *listenerEntry ? No need to make a copy of the objec.t

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 7 above)
Comment (costinm) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: Ah, I got it. Looks good.

File: pilot/pkg/model/service.go [hunk 9]
```diff
@@ -84,6 +86,9 @@ type Service struct {
 	// or use the passthrough model (i.e. proxy will forward the traffic to the network endpoint requested
 	// by the caller)
 	Resolution Resolution
+
```
Comment (costinm) [inline pilot/pkg/model/service.go]: Any reason to not add the "ID" ( == namespace + name for k8s, file path for file, etc ) ? It would really help - many times we get errors about a hostname or another, but in a multi-team environment it's very hard to know where that error comes from, since there is no good way to see where it was defined. 

Many setups restrict teams to namespaces - only admin can query across.

Comment (nmittler) [reply pilot/pkg/model/service.go]: @rshriram thoughts?

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 6 above)
Comment (nmittler) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: The default is a time of 0. It will work, and I have a test for it.

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 8 above)
Comment (nmittler) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: I was trying to avoid a bunch of small heap allocations.  The struct is quite small.

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 4 above)
Comment (nmittler) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: The function simplifies the call site

File: pilot/pkg/model/service.go (hunk 9 above)
Comment (rshriram) [reply pilot/pkg/model/service.go]: What would ID contain that hostname doesn't have? hostname is already name.namespace.svc.cluster.local. ID can only be populated if we know the source from which the service is being defined. But since these thigns are being sent to k8s API servers, we don't have the original file that defined it, unless we pull it out from API server (in which case I would call it as metadata, not ID). For CRDs, its the same issue.. One serviceEntry can become multiple services. So, multiple service entries will have same ID.

Comment (kyessenov) [reply pilot/pkg/model/service.go]: FWIW Mixer already defined the notion of service UID that is sent upstream in the control plane: https://github.com/istio/istio/blob/master/pilot/pkg/model/service.go#L371

Comment (rshriram) [reply pilot/pkg/model/service.go]: Yes, but the service attributes are populated only for k8s, used by RBAC.

Comment (kyessenov) [reply pilot/pkg/model/service.go]: They are used by mixer and and all mixer adapters...

**Comment by nmittler** [comment]
@costinm @rshriram I've been having trouble coming up with an integration test t hat recreates the original log message (I've been using the [simple example provided by @Stono](https://github.com/istio/istio/issues/5992#issuecomment-394821567).

It looks like #6402 changed the behavior of the listener conflict logic for TCP and I believe that this conflict will now no longer be detected/reported.  I suspect that the new code is incorrect and should be reverted, but I wanted to have another set of eyes take a look.  WDYT?
